ops/terraform/services/pipeline/modules/bfd_pipeline_scheduler/lambda_src/pipeline_scheduler.py

```diff
@@ -0,0 +1,411 @@
+import calendar
+import itertools
+import os
+import re
+import sys
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from enum import Enum
+from typing import Any, Optional
+from urllib.parse import unquote
+
+import boto3
+from botocore.config import Config
+
+SCALE_IN_FIVE_MINUTES_ACTION_NAME = "scale_in_in_five_minutes"
+SCALE_OUT_IMMEDIATELY_ACTION_NAME = "scale_out_immediately"
+SCALE_OUT_FUTURE_LOAD_ACTION_NAME_PREFIX = "scale_out_at_"
+
+REGION = os.environ.get("AWS_CURRENT_REGION", "us-east-1")
+BFD_ENVIRONMENT = os.environ.get("BFD_ENVIRONMENT", "")
+ETL_BUCKET_ID = os.environ.get("ETL_BUCKET_ID", "")
+PIPELINE_ASG_NAME = os.environ.get("PIPELINE_ASG_NAME", "")
+BOTO_CONFIG = Config(
+    region_name=REGION,
+    # Instructs boto3 to retry upto 10 times using an exponential backoff
+    retries={
+        "total_max_attempts": 10,
+        "mode": "adaptive",
+    },
+)
+
+try:
+    s3_resource = boto3.resource("s3", config=BOTO_CONFIG)
+    autoscaling_client = boto3.client("autoscaling", config=BOTO_CONFIG)
+    etl_bucket = s3_resource.Bucket(ETL_BUCKET_ID)
+except Exception as exc:
+    print(
+        f"Unrecoverable exception occurred when attempting to create boto3 clients/resources: {exc}"
+    )
+    sys.exit(0)
+
+
+class S3EventType(str, Enum):
+    """Represents the types of S3 events that this Lambda is invoked by and supports. The value of
+    each Enum is a substring that is matched for on the "eventName" property of an invocation
+    event"""
+
+    OBJECT_CREATED = "ObjectCreated"
+    OBJECT_REMOVED = "ObjectRemoved"
+
+
+class PipelineLoadType(str, Enum):
+    """Represents the possible types of data loads: either the data load is non-synthetic, meaning
+    that it contains production data and was placed within the root-level Incoming/Done folders of
+    the ETL bucket, or it is synthetic, meaning that it contains non-production, testing data and
+    was placed within the Incoming/Done folders within the Synthetic folder of the ETL bucket. The
+    value of each enum represents the name of the Incoming/Done folders' parent directory, with
+    empty string indicating that those paths have no parent"""
+
+    NON_SYNTHETIC = ""
+    SYNTHETIC = "Synthetic"
+
+
+class PipelineDataStatus(str, Enum):
+    """Represents the possible states of data: either data is available to load, or has been loaded
+    by the ETL pipeline. The value of each enum is the parent directory of the incoming file,
+    indicating status"""
+
+    INCOMING = "Incoming"
+    DONE = "Done"
+
+
+class RifFileType(str, Enum):
+    """Represents all of the possible RIF file types that can be loaded by the BFD ETL Pipeline. The
+    value of each enum is a specific substring that is used to match on each type of file
+    """
+
+    BENEFICIARY_HISTORY = "beneficiary_history"
```

Comment:
This file type will be removed from synthea (and I think as of cmac's work is no longer loaded in pipeline)

Reply:
Sounds good, I'll remove `BENEFICIARY_HISTORY` from this `Enum` then.